### PR TITLE
Allow any baud rate; not just a restricted set

### LINKED ISSUE #40: Only specific baud rates allowed, but chip supports any.
If you look in `serial_api.c` you can see that the nRF51822 code only allows a finite set of baud rates:

	static const int acceptedSpeeds[17][2] = {
		{1200, UART_BAUDRATE_BAUDRATE_Baud1200},
		{2400, UART_BAUDRATE_BAUDRATE_Baud2400},
		{4800, UART_BAUDRATE_BAUDRATE_Baud4800},
		{9600, UART_BAUDRATE_BAUDRATE_Baud9600},
		{14400, UART_BAUDRATE_BAUDRATE_Baud14400},
		{19200, UART_BAUDRATE_BAUDRATE_Baud19200},
		{28800, UART_BAUDRATE_BAUDRATE_Baud28800},
		{31250, (0x00800000UL) /* 31250 baud */},
		{38400, UART_BAUDRATE_BAUDRATE_Baud38400},
		{57600, UART_BAUDRATE_BAUDRATE_Baud57600},
		{76800, UART_BAUDRATE_BAUDRATE_Baud76800},
		{115200, UART_BAUDRATE_BAUDRATE_Baud115200},
		{230400, UART_BAUDRATE_BAUDRATE_Baud230400},
		{250000, UART_BAUDRATE_BAUDRATE_Baud250000},
		{460800, UART_BAUDRATE_BAUDRATE_Baud460800},
		{921600, UART_BAUDRATE_BAUDRATE_Baud921600},
		{1000000, UART_BAUDRATE_BAUDRATE_Baud1M}
	};

In fact the baud rate can be set to anything. Instead of using fixed constants, a formula should be used instead.

Old code:

	// serial_baud
	// set the baud rate, taking in to account the current SystemFrequency
	void serial_baud(serial_t *obj, int baudrate)
	{
		if (baudrate<=1200) {
			obj->uart->BAUDRATE = UART_BAUDRATE_BAUDRATE_Baud1200;
			return;
		}
	
		for (int i = 1; i<17; i++) {
			if (baudrate<acceptedSpeeds[i][0]) {
				obj->uart->BAUDRATE = acceptedSpeeds[i - 1][1];
				return;
			}
		}
		obj->uart->BAUDRATE = UART_BAUDRATE_BAUDRATE_Baud1M;
	}

Better code:

	// serial_baud
	// set the baud rate, taking in to account the current SystemFrequency
	void serial_baud(serial_t *obj, int baudrate)
	{
		// See https://devzone.nordicsemi.com/question/1181/uart-baudrate-register-values/#reply-1194
	
		// The correct value is baudrate * 2^32 / 16e6 = baudrate * 2^22 / 15625,
		// but that doesn't fit into 32 bits, and Cortex M0 has no hardware
		// divide, so instead I will do this possibly-over-optimised magic instead!
		// I have verified that this produces exactly the same values as the header,
		// except for 921600 (which is wrong in the header).
	
		uint32_t br = (baudrate << 8) + (baudrate << 4) - (baudrate << 1) - baudrate
				- (baudrate >> 1) - (baudrate >> 4) - (baudrate >> 9) - (baudrate >> 14)
				- (baudrate >> 16) - (baudrate >> 17) - (baudrate >> 18) - (baudrate >> 19);
	
		// Round it to 20 bits (see link above).
		br = (br + 0x800) & 0xFFFFF000;
		
		// Limit it to the valid range.
		if (br < 0x1000)
			br = 0x1000;
		if (br > 0x10000000)
			br = 0x10000000;
	
		obj->uart->BAUDRATE = br;
	}

I'll do a pull request for this.

### PR DESCRIPTION
**Pull request by Timmmm**
Fixes #40.

I'm pretty sure the rounding is unnecessary (i.e. the hardware rounds / truncates anyway), but I did it anyway so the values match the Nordic header.

### CODE REVIEW

**Comment by ciarmcom** [comment]
Automatic CI verification build not done, please verify manually.

**Comment by 0xc0170** [comment]
@emidttun @pan-

**Comment by ciarmcom** [comment]
Automatic CI verification build not done, please verify manually.

File: source/serial_api.c [hunk 1]
```diff
@@ -98,18 +79,28 @@ void serial_free(serial_t *obj)
...
+    // except for 921600 (which is wrong in the header).
+
+    uint32_t br = (baudrate << 8) + (baudrate << 4) - (baudrate << 1) - baudrate
+            - (baudrate >> 1) - (baudrate >> 4) - (baudrate >> 9) - (baudrate >> 14)
+            - (baudrate >> 16) - (baudrate >> 17) - (baudrate >> 18) - (baudrate >> 19);
+
+    // Round it to 20 bits (see link above).
+    br = (br + 0x800) & 0xFFFFF000;
+
+    // Limit it to the valid range.
```
Comment (emidttun) [inline source/serial_api.c]: Where is the valid range defined? From the nRF51 ref manual any 32-bit value seems to be accepted.

File: source/serial_api.c [hunk 2]
```diff
@@ -98,18 +79,28 @@ void serial_free(serial_t *obj)
...
+    // The correct value is baudrate * 2^32 / 16e6 = baudrate * 2^22 / 15625,
+    // but that doesn't fit into 32 bits, and Cortex M0 has no hardware
+    // divide, so instead I will do this possibly-over-optimised magic instead!
+    // I have verified that this produces exactly the same values as the header,
+    // except for 921600 (which is wrong in the header).
+
+    uint32_t br = (baudrate << 8) + (baudrate << 4) - (baudrate << 1) - baudrate
+            - (baudrate >> 1) - (baudrate >> 4) - (baudrate >> 9) - (baudrate >> 14)
+            - (baudrate >> 16) - (baudrate >> 17) - (baudrate >> 18) - (baudrate >> 19);
+
```
Comment (emidttun) [inline source/serial_api.c]: I would really like to have test cases for this formula, verifying that the values match the ones in the Nordic ref manual. But if you have verified manually I would say that is good enough.

**Comment by Timmmm** [comment]
Here is the code I used to test it. It's Matlab/Octave rather than C, but it is the same maths.

```
function code = ToBaudCode(br)

code = bitshift(br, 8, 32) ...
+ bitshift(br, 4, 32)...
 - bitshift(br, 1, 32)...
 - br...
 - bitshift(br, -1, 32)...
 - bitshift(br, -4, 32)...
 - bitshift(br, -9, 32)...
 - bitshift(br, -14, 32)...
 - bitshift(br, -16, 32)...
 - bitshift(br, -17, 32)...
 - bitshift(br, -18, 32)...
 - bitshift(br, -19, 32);
 
 % Round it
 code = bitand(code + 0x800, 0xFFFFF000);
 
 end
```

Test:

```
bauds = [1200 2400 4800 9600 14400 19200 28800 38400 57600 76800 115200 230400 250000 460800 921600 1000000];
for ff = bauds
    dec2hex(ToBaudCode(ff))
end
```

Output:

```
ans = 4F000
ans = 9D000
ans = 13B000
ans = 275000
ans = 3B0000
ans = 4EA000
ans = 75F000
ans = 9D5000
ans = EBF000
ans = 13A9000
ans = 1D7E000
ans = 3AFB000
ans = 4000000
ans = 75F7000
ans = EBEE000
ans = 10000000
```

Correct values:
```
#define UART_BAUDRATE_BAUDRATE_Baud1200 (0x0004F000UL) /*!< 1200 baud. */
#define UART_BAUDRATE_BAUDRATE_Baud2400 (0x0009D000UL) /*!< 2400 baud. */
#define UART_BAUDRATE_BAUDRATE_Baud4800 (0x0013B000UL) /*!< 4800 baud. */
#define UART_BAUDRATE_BAUDRATE_Baud9600 (0x00275000UL) /*!< 9600 baud. */
#define UART_BAUDRATE_BAUDRATE_Baud14400 (0x003B0000UL) /*!< 14400 baud. */
#define UART_BAUDRATE_BAUDRATE_Baud19200 (0x004EA000UL) /*!< 19200 baud. */
#define UART_BAUDRATE_BAUDRATE_Baud28800 (0x0075F000UL) /*!< 28800 baud. */
#define UART_BAUDRATE_BAUDRATE_Baud38400 (0x009D5000UL) /*!< 38400 baud. */
#define UART_BAUDRATE_BAUDRATE_Baud57600 (0x00EBF000UL) /*!< 57600 baud. */
#define UART_BAUDRATE_BAUDRATE_Baud76800 (0x013A9000UL) /*!< 76800 baud. */
#define UART_BAUDRATE_BAUDRATE_Baud115200 (0x01D7E000UL) /*!< 115200 baud. */
#define UART_BAUDRATE_BAUDRATE_Baud230400 (0x03AFB000UL) /*!< 230400 baud. */
#define UART_BAUDRATE_BAUDRATE_Baud250000 (0x04000000UL) /*!< 250000 baud. */
#define UART_BAUDRATE_BAUDRATE_Baud460800 (0x075F7000UL) /*!< 460800 baud. */
#define UART_BAUDRATE_BAUDRATE_Baud921600 (0x0EBED000UL) /*!< 921600 baud. */
#define UART_BAUDRATE_BAUDRATE_Baud1M (0x10000000UL) /*!< 1M baud. */
```

Looks like the 921600 value has been rounded in the latest code. My value is very slightly different, but it shouldn't matter in practice (it's a 0.002% difference).

It seems pretty clear from the datasheet to me that the maximum rate is 1 Mbaud, especially as 1 Mbaud gets a special value (0x10000000). The minimum rate is due to the rounding (I don't think anyone will want to go as low as 15 baud anyway!).

**Pull request closed without merging**